packages/protocol/contracts/team/airdrop/MerkleClaimable.sol

```diff
@@ -0,0 +1,51 @@
+// SPDX-License-Identifier: MIT
+//  _____     _ _         _         _
+// |_   _|_ _(_) |_____  | |   __ _| |__ ___
+//   | |/ _` | | / / _ \ | |__/ _` | '_ (_-<
+//   |_|\__,_|_|_\_\___/ |____\__,_|_.__/__/
+
+pragma solidity ^0.8.20;
+
+import { OwnableUpgradeable } from
+    "lib/openzeppelin-contracts-upgradeable/contracts/access/OwnableUpgradeable.sol";
+import { MerkleProofUpgradeable } from
+    "lib/openzeppelin-contracts-upgradeable/contracts/utils/cryptography/MerkleProofUpgradeable.sol";
+
+import { Proxied } from "../../common/Proxied.sol";
+
+/// @title MerkleClaimable
+/// Contract for managing Taiko token airdrop for eligible users
+// TODO(dani): add claimStart and claimEnd timestamp so claim() can only be
+// called between these two timestamps.
+// Better to add non-reentrance guard to claim().
+
+abstract contract MerkleClaimable is OwnableUpgradeable {
+    mapping(bytes32 => bool) public isClaimed;
+    bytes32 public merkleRoot;
+
+    event Claimed(bytes32 hash);
+
+    error CLAIMED_ALREADY();
+    error INVALID_PROOF();
+
+    function claim(bytes calldata data, bytes32[] calldata proof) external {
+        bytes32 hash = keccak256(abi.encode("CLAIM_TAIKO_AIRDROP", data));
```

Comment:
been playing around with it for a while now (testing the off-chain, on-chain scripts) now, and figured it is not going to work out properly, i'd suggest something like this:

```
(address user, uint256 amount) = abi.decode(data, (address, uint256));

        bytes32 hash = keccak256(abi.encode("CLAIM_TAIKO_AIRDROP", user, amount));
```

The reason is, calldata has some 'marker' appended to the data itself.

So if i do this:

        `console2.logBytes(abi.encode("CLAIM_TAIKO_AIRDROP", data));`

The result is this:
`0x000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000000000000000000000000000000000013434c41494d5f5441494b4f5f41495244524f500000000000000000000000000000000000000000000000000000000000000000000000000000000000000000400000000000000000000000007e5f4552091a69125d5dfcb7b8c2659029395bdf0000000000000000000000000000000000000000000000000000000000000064` 

But if i decode the value (as proposed above):
`0x00000000000000000000000000000000000000000000000000000000000000600000000000000000000000007e5f4552091a69125d5dfcb7b8c2659029395bdf00000000000000000000000000000000000000000000000000000000000000640000000000000000000000000000000000000000000000000000000000000013434c41494d5f5441494b4f5f41495244524f5000000000000000000000000000`

So it is more easy to create the raw data off-chain if we dont need to consider solidity encoding structure.